### PR TITLE
Add DbException.IsTransient property

### LINKED ISSUE #34817: Database-agnostic way to detect transient database errors
A recurring theme for database programming has been building automated retry policies, but System.Data doesn't currently allow database drivers to expose whether a database error can be expected to be transient or not. This leads to database-specific logic for detecting transient exceptions being built again and again outside of the driver.

# API Proposal

```c#
class DbException
{
    public virtual bool IsTransient => false;
}
```

# Notes

* IsTransient should be *optimistic*, i.e. return true in all cases where a simple retry of the operation - without any other change - *may* be successful. We prefer to err on the side of transience, since the downside is only a few extra retries.
* In at least some cases, if an operation transiently fails within a transaction, the entire transaction must be rolled back and retried (including all previous operations). This would be out of scope of this proposal: DbException would only inform the consumers that the exception is transient, but the actual handling (rolling back, reissuing previous commands) would be the consumer's responsibility.

# Resources and examples

* EF Core's [SQL Server transient exception detector](https://github.com/dotnet/efcore/blob/bde2b140d6f4cf94d6d1285d402941e20193ec60/src/EFCore.SqlServer/Storage/Internal/SqlServerTransientExceptionDetector.cs).
* Npgsql's [NpgsqlException.IsTransient property](https://github.com/npgsql/npgsql/blob/9bd3fff5ff6930011cf64521adaebb29f8363f52/src/Npgsql/PostgresException.cs#L162), which is likely to resemble the API for this proposal.
* [SqlClient's internal retry mechanism, important especially for Azure SQL](https://docs.microsoft.com/en-us/azure/sql-database/sql-database-connectivity-issues#net-sqlconnection-parameters-for-connection-retry). See also [this conversation](https://github.com/dotnet/SqlClient/issues/416) on possibly evolving that mechanism.

# Old notes on SupportsTransienceDetection (removed from proposal)

* SupportsTransienceDetection can be used to signal that IsTransient has been implemented by a driver, and can be expected to be true where relevant. Note that in general it should be fine to just check IsTransient directly, as it would default to false in any case.
* It's a bit problematic for  to be on DbException:
    * This makes it hard to do a simple check on startup (one needs an actual DbException instance thrown from the driver).
    * This feature flag would ideally be in a general metadata/feature discovery facility (see #28763), but we don't have that at the moment.
    * We could put this flag on DbProviderFactory, but that would turn it into a general place for feature flags as opposed to a factory (current flags such as [CanCreateDataAdapter](https://docs.microsoft.com/en-us/dotnet/api/system.data.common.dbproviderfactory.cancreatedataadapter?view=netcore-3.1#System_Data_Common_DbProviderFactory_CanCreateDataAdapter) are about factory methods on the type, so much more appropriate).
    * As the problem isn't very acute, having this property on DbException is probably OK.

/cc @David-Engel @cheenamalhotra @bgrainger @bricelam @ajcvickers @stephentoub @terrajobst @mgravell @FransBouma

# Edit history

| Date | Modification |
| ----- | ------------ |
| 2020-04-10 | API proposal
| 2020-04-18 | Removed SupportsTransienceDetection

### PR DESCRIPTION
**Pull request by rgarrison12345**
Fixes https://github.com/dotnet/runtime/issues/34817

### CODE REVIEW

File: src/libraries/System.Data.Common/src/System/Data/Common/DbException.cs [hunk 1]
```diff
@@ -8,6 +8,8 @@ namespace System.Data.Common
     [System.Runtime.CompilerServices.TypeForwardedFrom("System.Data, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")]
     public abstract class DbException : System.Runtime.InteropServices.ExternalException
     {
+        public virtual bool IsTransient => false;
```
Comment (jkotas) [inline src/libraries/System.Data.Common/src/System/Data/Common/DbException.cs:11]: Nit: We typically put properties after the constructors in the implementation.

**Comment by jkotas** [comment]
Could you please also add a test?

**Comment by rgarrison12345** [comment]
Hello @jkotas, I'm aware that for a new pull request Microsoft would like to see a test for the functionality. 

However, this is adding a property to the DbException class with a default implementation.  The idea is providers like Microsoft.Data.SqlClient, MySqlConnector, NpgSql will actually implement that property.  With that in mind I didn't see the value of a test because the derived class is expected to set that property.  If there is something specific you have in mind I can do the test

File: src/libraries/System.Data.Common/src/System/Data/Common/DbException.cs (hunk 1 above)
Comment (rgarrison12345) [reply src/libraries/System.Data.Common/src/System/Data/Common/DbException.cs:11]: Hello @jkotas, do I need to submit a separate pull request to update the styling of the code file to match Microsoft expectations?

**Comment by jkotas** [comment]
The test can be very simple: Just call the base class property and check that it returns false.

File: src/libraries/System.Data.Common/src/System/Data/Common/DbException.cs (hunk 1 above)
Comment (jkotas) [reply src/libraries/System.Data.Common/src/System/Data/Common/DbException.cs:11]: You can push a new commit into this PR to fix this nit, no need to create a new PR.

**Comment by jkotas** [comment]
Could you please resolve the conflict?

File: src/libraries/System.Data.Common/tests/System/Data/Common/DbExceptionTests.cs [hunk 2]
```diff
@@ -16,7 +16,13 @@ public void Ctor_Parameterless_UsesDefaults()
             Assert.Null(defaultException.InnerException);
             Assert.Equal(-2147467259, defaultException.ErrorCode);
         }
+        [Fact]
```
Comment (stephentoub) [inline src/libraries/System.Data.Common/tests/System/Data/Common/DbExceptionTests.cs:19]:
```suggestion

        [Fact]
```

File: src/libraries/System.Data.Common/tests/System/Data/Common/DbExceptionTests.cs [hunk 3]
```diff
@@ -16,7 +16,13 @@ public void Ctor_Parameterless_UsesDefaults()
...
+        public void Ctor_IsTransient_IsFalse()
+        {
+            var defaultException = new CustomDbException();
 
+            Assert.False(defaultException.IsTransient);
+        }
```
Comment (stephentoub) [inline src/libraries/System.Data.Common/tests/System/Data/Common/DbExceptionTests.cs:25]:
```suggestion
        }

```

**Review by stephentoub — APPROVED** [review]

**Comment by roji** [comment]
@rgarrison12345 thanks for this! Can you please squash your commits and rebase them so it's a single, clean commit on top of the latest master branch commits (no merge commits)?

Note that there's also #35601 for adding SqlState to DbException. There was also a proposal in the [design discussion](https://www.youtube.com/watch?v=5w9Cbgr1HJY&t=0h25m54s) to provide a default implementation of IsTransient by looking at SqlState, and identifying some specific values which are inherently transient. This doesn't have to be done right away though.

File: src/libraries/System.Data.Common/ref/System.Data.Common.cs [hunk 4]
```diff
@@ -2197,8 +2197,12 @@ public partial class DbEnumerator : System.Collections.IEnumerator
 #nullable enable
     public abstract partial class DbException : System.Runtime.InteropServices.ExternalException
     {
+        public virtual bool IsTransient { get { throw null; } }
         protected DbException() { }
         protected DbException(System.Runtime.Serialization.SerializationInfo info, System.Runtime.Serialization.StreamingContext context) { }
+        protected DbException(string message) { }
```
Comment (roji) [inline src/libraries/System.Data.Common/ref/System.Data.Common.cs:2203]: Looks like this slippted in by mistake, please remove.

**Review by roji — COMMENTED** [review]
Please check the spurious DbException constructors introduced here. It's also best to squash these 13 commits into a single one on top of master (no merge commits necessary).

**Comment by roji** [comment]
@rgarrison12345 any reason for closing this? If you intend to open a new PR for this, please reopen this one instead, and force-push to its branch (a new PR would require new approval etc.). Otherwise please let me know and I'll do this functionality.

**Comment by rgarrison12345** [comment]
> 
> 
> @rgarrison12345 any reason for closing this? If you intend to open a new PR for this, please reopen this one instead, and force-push to its branch (a new PR would require new approval etc.). Otherwise please let me know and I'll do this functionality.

Hello @roji, I'll admit I'm fairly new with the Git workflow.  I'm more familiar with TFS.  When I did the pull request originally I thought I had latest, but it's apparent I didn't and that was causing some issues. I was closing this to open a new one to start over fresh.

Can you do this functionality?  I want to contribute, I think I need to learn more about git in the mean time.

**Comment by roji** [comment]
Sure, I'll soon submit a PR for both this and #35601. Git is indeed non-trivial for newcomers, but it's well-worth the investment to learn it! Hope to see more contributions from you soon.

**Comment by roji** [comment]
Replaced by #39157

**Pull request closed without merging**